vispy/visuals/colorbar.py

```diff
@@ -0,0 +1,330 @@
+        on the side of the colorbar.
+    orientation : {'horizontal', 'vertical'}
+        the orientation of the colorbar, used for coloring
+
+            * 'horizontal': the color is applied from left to right,
+              with minimum corresponding to left and maximum to right
+            * 'vertical': the color is applied from bottom to top,
+              with mimumum corresponding to bottom and maximum to top
+    border_width : float (in px)
+        The width of the border the colormap should have. This measurement
+        is given in pixels
+    border_color : string | vispy.color.Color
+        The color of the border of the colormap. This can either be a
+        string as the color's name or an actual instace of a vipy.color.Color
+    """
+
+    def __init__(self, center_pos, halfdim,
+                 label,
+                 cmap,
+                 clim=(0.0, 1.0),
+                 orientation="horizontal",
+                 border_width=1.0,
+                 border_color="black",
+                 **kwargs):
+
+        super(ColorBarVisual, self).__init__(**kwargs)
+
+        self._cmap = get_colormap(cmap)
+        self._clim = clim
+        self._center_pos = center_pos
+        self._halfdim = halfdim
+        self._orientation = orientation
+        self._border_width = border_width
+        self._border_color = border_color
+
+        # initialize the labels and ticks that will be
+        # positioned with a call to regenerate
+        self._label = TextVisual(label, anchor_y="top")
+        self._ticks = []
+        self._ticks.append(TextVisual(str(self._clim[0]), anchor_y="top"))
+        self._ticks.append(TextVisual(str(self._clim[1]), anchor_y="top"))
+
+        # setup border rendering
+        self._border_color = Color(border_color)
+        self._border_program = ModularProgram(VERT_SHADER_BORDER,
+                                              FRAG_SHADER_BORDER)
+
+        adjust_dir = np.array([[-1, -1], [1, -1], [1, 1],
+                              [-1, -1], [1, 1], [-1, 1]],
+                              dtype=np.float32)
+
+        self._border_program['a_adjust_dir'] = adjust_dir.astype(np.float32)
+
+        # setup the right program shader based on color
+        if orientation == "horizontal":
+            self._program = ModularProgram(VERT_SHADER, FRAG_SHADER_HORIZONTAL)
+
+        elif orientation == "vertical":
+            self._program = ModularProgram(VERT_SHADER, FRAG_SHADER_VERTICAL)
+        else:
+            raise ColorBarVisual._get_orientation_error(self._orientation)
+
+        tex_coords = np.array([[0, 0], [1, 0], [1, 1],
+                              [0, 0], [1, 1], [0, 1]],
+                              dtype=np.float32)
+
+        self._program.frag['color_transform'] = Function(self._cmap.glsl_map)
+        self._program['a_texcoord'] = tex_coords.astype(np.float32)
+
+        self.regenerate()
+
+    def regenerate(self):
```

Comment:
Ah, yes. It should have been private in the first place.

Well, whenever you need to recreate shader data or recreate the label or
ticks or something, it's called.

So, for example, when you change the border width, the positions of the
labels need to be regenerated to not overlap with the new border.

Sent from my phone, please excuse any typos
On Jun 9, 2015 7:26 AM, "Eric Larson" <notifications@github.com> wrote:

> In vispy/visuals/colorbar.py
> <https://github.com/vispy/vispy/pull/979#discussion_r31977501>:
>
> > +
> > +        elif orientation == "vertical":
> > +            self._program = ModularProgram(VERT_SHADER, FRAG_SHADER_VERTICAL)
> > +        else:
> > +            raise ColorBarVisual._get_orientation_error(self._orientation)
> > +
> > +        tex_coords = np.array([[0, 0], [1, 0], [1, 1],
> > +                              [0, 0], [1, 1], [0, 1]],
> > +                              dtype=np.float32)
> > +
> > +        self._program.frag['color_transform'] = Function(self._cmap.glsl_map)
> > +        self._program['a_texcoord'] = tex_coords.astype(np.float32)
> > +
> > +        self.regenerate()
> > +
> > +    def regenerate(self):
>
> when does it need to be called? it should probably be private for now
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/vispy/vispy/pull/979/files#r31977501>.
>